### PR TITLE
Token and CRUD operations of user and projects

### PR DESCRIPTION
**Pull request by lhamu**
Added the JWT access and refresh tokens and the CRUD operations for the user.

### CODE REVIEW

File: src/config/passport.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+module.exports = function(passport) {
+  passport.use(new GoogleStrategy({
+    clientID: clientInfo.googleAuth.clientID,
+    clientSecret: clientInfo.googleAuth.clientSecret,
+    callbackURL: clientInfo.googleAuth.callbackURL
+  },
+  
+  function(accessToken, refreshToken, profile, done) {
+    let data = {
```
Comment (badyakarprabin) [inline src/config/passport.js:23]: Can we do this here?  
```
const { id,  displayName, emails,  _json} = profile ;

let data = {
  'id': id,
  'name': displayName,
  'email': emails[0].values,
  'accessToken': accessToken,
  'image': _json.image.url
}
```

**Comment by badyakarprabin** [comment]
@lhamu Things to remember before creating PR:

- Remove unwanted codes (commented code).
- Write proper JS Docs.

Useful Tips:
[Auto code formatting while saving
](https://stackoverflow.com/questions/12496669/code-reformatting-on-save-in-phpstorm)
[Contributing guidelines](https://github.com/leapfrogtechnology/chill/blob/master/CONTRIBUTING.md)

File: src/app.js [hunk 2]
```diff
@@ -13,6 +13,12 @@ import * as errorHandler from './middlewares/errorHandler';
 
 const app = express();
 
+import passport from 'passport';
```
Comment (deepsadhi) [inline src/app.js:16]: Can we move these lines to top?

File: src/config/passport.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import passport from 'passport';
+import GooglePassport from 'passport-google-oauth';
+import * as clientInfo from './auth';
```
Comment (deepsadhi) [inline src/config/passport.js:3]: Reorder statement with by statement length

File: src/app.js [hunk 4]
```diff
@@ -13,6 +13,12 @@ import * as errorHandler from './middlewares/errorHandler';
...
+import passportConfig from './config/passport';
+
+
```
Comment (deepsadhi) [inline src/app.js:19]: Remove line

File: src/config/passport.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// import * as User from '../models/users';
```
Comment (deepsadhi) [inline src/config/passport.js:4]: If this comment is not required lets remove it

File: src/config/passport.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+let GoogleStrategy = GooglePassport.OAuth2Strategy;
+/*
```
Comment (deepsadhi) [inline src/config/passport.js:7]: Indent comments properly

File: src/config/passport.js (hunk 1 above)
Comment (deepsadhi) [reply src/config/passport.js:23]:
```js
let data = {
  id,
  accessToken,
  name: displayName,
  image: _json.image.url,
  email: emails[0].values
}
```

File: src/config/passport.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      'email': profile.emails[0].value,
+      'accessToken': accessToken,
+      'image': profile._json.image.url
+    };
+        
+    return done(null, data);
+  }
+));
+
+  passport.serializeUser(function(user, done) {
```
Comment (deepsadhi) [inline src/config/passport.js:35]:
```js
passport.serializeUser((user, done) => done(null, user));
```

File: src/controllers/profile.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import { Router } from 'express';
+// import * as authenticate from '../middlewares/authenticate';
```
Comment (deepsadhi) [inline src/controllers/profile.js:2]: Let's remove comment if not required

File: src/controllers/profile.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+let router = Router();
+
+// router.post('/', (request, response)=>{
```
Comment (deepsadhi) [inline src/controllers/profile.js:6]: Let's remove comment if not required

File: src/controllers/profile.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+//   console.log('login');
+//   let data=request.body;
+//   console.log(data);
+//   response.json({data});
+// });
+
+router.get('/', (request, response)=>{
```
Comment (deepsadhi) [inline src/controllers/profile.js:13]: * Proper spacing and indentation
* Use return in same line if there is only one statement
```js
router.get('/', (request, response) => response.json('User profile'));
```

File: src/controllers/project.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
+// Project Controller
+import * as projectService from '../services/project';
+import * as userProjectService from '../services/userProject';
+import jwt from 'jsonwebtoken';
```
Comment (deepsadhi) [inline src/controllers/project.js:4]: Import npm packages at top and then import custom code separated by a blank line
```js
import jwt from 'jsonwebtoken';

import * as projectService from '../services/project';
import * as userProjectService from '../services/userProject';
```

File: src/controllers/project.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+/*
```
Comment (deepsadhi) [inline src/controllers/project.js:6]: Indent properly

File: src/controllers/project.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    Create a Project
+    Returns the project json
+*/
+export function create( req, res, next ) {
```
Comment (deepsadhi) [inline src/controllers/project.js:10]:
```js
export function create(req, res, next) {
```

File: src/controllers/project.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  projectService.create( req.body )
```
Comment (deepsadhi) [inline src/controllers/project.js:11]: Same here

File: src/controllers/project.js [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    .then(data =>{
+            // Call userproject service to map user and project
```
Comment (deepsadhi) [inline src/controllers/project.js:13]: Indent properly

File: src/controllers/project.js [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+      let head = req.headers.authorization.split(' ')[1];
+      let value = jwt.decode(head);
+      let userProjectData = {
+        projectId: data.id,
+        userId: value.userId
+      };
+      // res.json(userProjectData);
```
Comment (deepsadhi) [inline src/controllers/project.js:20]: Remove line if not required

File: src/controllers/project.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+      userProjectService.create(userProjectData)
+        .then(userprojectreturn =>{
```
Comment (deepsadhi) [inline src/controllers/project.js:23]: Proper spacing

File: src/controllers/project.js [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * @param {Object} req
+ * @param {Object} res
+ * @param {Object} next
+ */
+export function deleteProject( req , res , next ) {
+  let head = req.headers.authorization.split(' ')[1];
+  let value = jwt.decode(head);
+  
+  projectService.deleteProject(value.userId, req.params.projectid )
```
Comment (deepsadhi) [inline src/controllers/project.js:80]:
```js
projectService.deleteProject(value.userId, req.params.projectid)
```

File: src/controllers/project.js [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * @param {Object} req
+ * @param {Object} res
+ * @param {Object} next
+ */
+export function updateProject( req , res , next ) {
+  let head = req.headers.authorization.split(' ')[1];
+  let value = jwt.decode(head);
+
+  projectService.updateProject( value.userId, req.params.projectid , req.body )
```
Comment (deepsadhi) [inline src/controllers/project.js:96]:
```js
projectService.updateProject(value.userId, req.params.projectid, req.body)
```

File: src/controllers/users.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
+import * as userService from '../services/user';
+
+/*
+create a user
```
Comment (deepsadhi) [inline src/controllers/users.js:4]: Indentation

File: src/controllers/users.js [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+*/
+export function create( req, res, next ) {
+  userService.createUser(req.user)
+    .then(data => res.json(data))
+    .catch(err => next(err));
+}
+
+/*
+fetch a user from id
```
Comment (deepsadhi) [inline src/controllers/users.js:13]: Same here

File: src/controllers/users.js [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+*/
+export function get( req, res, next) {
+    // console.log('hello');
```
Comment (deepsadhi) [inline src/controllers/users.js:16]: Let 's remove console.log

File: src/middlewares/checkToken.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
+import * as tokenServices from '../services/token';
```
Comment (deepsadhi) [inline src/middlewares/checkToken.js:1]: Import order

File: src/queries/project.js [hunk 24]
```diff
@@ -0,0 +1,38 @@
+// Project Queries
+// Select all Projects of a user
+export const FETCH_All_PROJECTS = (`
+    select *
```
Comment (deepsadhi) [inline src/queries/project.js:4]: * Use of SQL keywords in UPPERCASE
* Do we require bracket `()`?
```js
export const FETCH_All_PROJECTS = `
  SELECT * 
  FROM users u  
  LEFT JOIN user_project up ON up.user_id = u.id 
  LEFT JOIN projects p ON up.project_id = p.id 
  WHERE u.id = ?;
`
```

File: src/services/user.js [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+export async function loginOrSignUp(data) {
+  try {
+
+    let email = data.email;
+    let userInfo = await fetchByEmail(email);
+
+    console.log(userInfo);
+
+    if (userInfo) {
+      let accessToken = generateTokens.generateToken(userInfo.id, 'RESTFULAPI', 300);
```
Comment (deepsadhi) [inline src/services/user.js:41]: Remove magic numbers. Use constants instead.

**Review by deepsadhi — CHANGES_REQUESTED** [review]
1. Import npm packages at top and custom packages separated by blank line
2. Indent code properly with spacing  
3. Use of doc block 
4. Remove use of return keyword if return statement is of single line [  func1(obj => func2(obj2))  ] 
5. Remove console.log before code is pushed
6. Remove ' from object key. Eg. { name: 'ABC' } instead of { 'name': 'ABC' }
7. If key and variable is same of object use { name } instead of { name: name }
8. Order statements by statement length order
9. Follow same if syntax indentation by everyone
10. Remove comments if not required
11. Do not leave multiple lines, one line is enough
12. Write all SQL keywords in UPPERCASE
13. Remove magic numbers. Use constant instead of it
14. Use of single quote ' instead of double quote "

**Comment by romitamgai** [comment]
This is old pull request. Closing this. Thanks for the comments though @deepsadhi

**Pull request closed without merging**